R/PipeOpDateFeatures.R

```diff
@@ -0,0 +1,226 @@
+        month = TRUE, week_of_year = TRUE, day_of_year = TRUE, day_of_month = TRUE,
+        day_of_week = TRUE, hour = TRUE, minute = TRUE, second = TRUE, is_day = TRUE)
+      super$initialize(id = id, param_set = ps, param_vals = param_vals)
+    },
+
+    select_cols = function(task) {
+      task$feature_types[get("type") == "POSIXct", get("id")]
+    },
+
+    transform_dt = function(dt, levels) {
+      features = names(which(unlist(self$param_set$get_values(tags = "datepart"))))
+      # early exit if no features are to be computed
+      if (length(features) == 0L) {
+        return(dt)
+      }
+
+      # special handling of year because this is needed for day_of_year and day_of_month
+      drop_year = "year" %nin% features
+      features = unique(c("year", features))
+
+      cyclic_features = features[features %in% c("month", "week_of_year", "day_of_year",
+        "day_of_month", "day_of_week", "hour", "minute", "second")]
+
+      date_vars = colnames(copy(dt))
+
+      for (date_var in date_vars) {
+        dt[, paste0(date_var, ".", features) :=
+          compute_date_features(get(date_var), features = features)]
+      }
+
+      # if cyclic = TRUE for month, week_of_year, day_of_year, day_of_month, day_of_week, hour,
+      # minute and second, two columns are additionally added, each consisting of their sine and
+      # cosine transformation of in general 2 * pi * x / max_x + 1 (x starting from 0)
+      if (self$param_set$values$cyclic && (length(cyclic_features) > 0L)) {
+        for (date_var in date_vars) {
+          dt[, paste0(date_var, ".", rep(cyclic_features, each = 2L), "_", c("sin", "cos")) :=
+            compute_cyclic_date_features(.SD, features = cyclic_features, date_var = date_var)]
+        }
+      }
+
+      if (self$param_set$values$keep_date_var == FALSE) {
+        dt[, (date_vars) := NULL]
+      }
+
+      if (drop_year) {
+        dt[, paste0(date_vars, ".", "year") := NULL]
+      }
+
+      dt
+    }
+  )
+)
+
+mlr_pipeops$add("datefeatures", PipeOpDateFeatures)
+
+# helper function to compute date features of a vector of dates
+compute_date_features = function(dates, features) {
+  lapply(features, FUN = function(feature) {
+    formatting = switch(feature,
+      year = "%Y",
+      month = "%m",
+      week_of_year = "%U", # starting on Sunday
+      day_of_year = "%j",
+      day_of_month = "%d",
+      day_of_week = "%w", # 0 = Sunday
+      hour = "%H",
+      minute = "%M",
+      second = "%S",
+      is_day = "%H")
+    if (feature == "is_day") {
+      hours = as.numeric(format(dates, formatting))
+      return((6L <= hours) & (hours <= 20L)) # early exit
+    }
+    as.numeric(format(dates, formatting))
+  })
+}
+
+# helper function to compute cyclic date features of date features, i.e.,
+# sine and cosine transformations of 2 * pi * x / max_x + 1
+compute_cyclic_date_features = function(date_features, features, date_var) {
```

Comment:
is there a more elegant way to get to the date values instead of using `date_features[[paste0(date_var, ".", XXX)]]` all the time? Maybe just give the `date_var.feature` column as an argument directly?